### PR TITLE
build: enable strictPropertyInitialization

### PR DESCRIPTION
**Pull request by Splaktar**
- fix related build errors
- complete `this._destroyed` in `ngOnDestroy()`
- replace use of deprecated `DomPortalHost` with `DomPortalOutlet`

Builds on PR https://github.com/angular/material.angular.io/pull/688

### CODE REVIEW

**Comment by Splaktar** [comment]
@crisbeto can you please take a look at this [Circle CI](https://circleci.com/gh/angular/material.angular.io/441) error? It seems to be related to the [Portal Changes in 9.0.0](https://github.com/angular/components/pull/16708).

Should I file a bug here or is there some misconfiguration in this PR?

File: src/app/shared/doc-viewer/doc-viewer.ts [hunk 1]
```diff
@@ -19,16 +19,16 @@ import {Subscription} from 'rxjs';
 import {take} from 'rxjs/operators';
 import {ExampleViewer} from '../example-viewer/example-viewer';
 import {HeaderLink} from './header-link';
+import {DomPortalOutlet} from '@angular/cdk/portal/dom-portal-outlet';
```
Comment (devversion) [inline src/app/shared/doc-viewer/doc-viewer.ts]: As mentioned on Slack. This import seems the culprit. Any reason why you use a deep import?

Comment (Splaktar) [reply src/app/shared/doc-viewer/doc-viewer.ts]: That's what WebStorm defaulted to when auto importing 😞

**Comment by Splaktar** [comment]
I just found that this branch/PR were passing on CircleCI, but failing to build locally. It appears that `ng test` doesn't expose Ivy's Template Type Checking errors. However, Template Type Checking does break the build. So I've added a build task to CI to catch this issue.

Additionally, it also runs `build:content` to ensure that the docs-content can be synced as we've had a couple issues this year where that was broken and we weren't aware of it immediately.

**Comment by Splaktar** [comment]
OK, the issues that I'm hitting were introduced in Angular `9.0.0-rc.5` as part of PR https://github.com/angular/angular/pull/33997 (specifically https://github.com/angular/angular/pull/33997/commits/66e8ed4ca8c7bd140e7dd3de0493e21382d2d755) which was fixing https://github.com/angular/angular/issues/31556.

I'm investigating how to solve them, but I don't feel like my current solution is solid, so I'm continuing to investigate.

**Comment by Splaktar** [comment]
@crisbeto or @JoostK do you have any guidance here?

### Template
```html
<ng-container *ngIf="componentViewer.componentDocItem | async; let docItem">
  <span class="cdk-visually-hidden" tabindex="-1" #initialFocusTarget>
    Examples for {{docItem.id}}
  </span>
  <example-viewer *ngFor="let example of docItem.examples"
                  [example]="example"></example-viewer>
</ng-container>
```
There are errors for
```
src/app/pages/component-viewer/component-examples.html:3:20 - error TS2531: Object is possibly 'null'.

3     Examples for {{docItem.id}}
                     ~~~~~~~~~~

  src/app/pages/component-viewer/component-viewer.ts:132:16
    132   templateUrl: './component-examples.html',
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component ComponentExamples.
src/app/pages/component-viewer/component-examples.html:5:34 - error TS2531: Object is possibly 'null'.

5   <example-viewer *ngFor="let example of docItem.examples" [example]="example"></example-viewer>
                                   ~~~~~~~~~~~~~~~~

  src/app/pages/component-viewer/component-viewer.ts:132:16
    132   templateUrl: './component-examples.html',
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component ComponentExamples.
```

# Solutions?

It's complaining about `docItem.id` and `docItem.examples` where `docItem` might be `null` in these cases. But it seems to be ignoring the fact that this template won't be rendered if `docItem` is `null`.

The tests in PR https://github.com/angular/angular/pull/33997 don't cover the use case where the Async Pipe is used.

How do you recommend developers solve this very common use case?

**Comment by JoostK** [comment]
@Splaktar ah, this is interesting. Using `as` the way we tested is reflects the `NgIfContext.ngIf` property, for which the template type checker has applied the appropriate context guard expression due to the presence of `NgIf.ngContextGuard_ngIf`.

When used with a let assignment to `$implicit`, however, the context guard for `ngIf` does not narrow reads of `$implicit`.

I'll have a look to see if there's something that can be done here.

**Comment by Splaktar** [comment]
@JoostK I tried the following and got the same errors:
```html
<ng-container *ngIf="componentViewer.componentDocItem | async as docItem">
```

**Comment by Splaktar** [comment]
@JoostK here is the [workaround](https://github.com/angular/material.angular.io/commit/f813cc6d0c12bc52dab5237922403fcde4108e72) that I had to apply to get our repo building again after the breaking change in Angular `9.0.0-rc.5`. Do you have an open GitHub issue to track this or should I open one?

**Comment by JoostK** [comment]
@Splaktar there isn't any issue at the moment, so please open one.

In the meantime I am working on a PR to resolve this issue, but I'm not fond of the approach quite yet.

The main reason as to why template variables are not properly narrowed is because [the "template guard" for the `ngIf` input](https://github.com/angular/angular/blob/cca26166376d4920f5905b168e70ea2e8d70da77/packages/common/src/directives/ng_if.ts#L227) is only regarding the input binding expression itself. When using template variables, however, the value is taken from `NgIfContext` which is _not_ narrowed as it's currently not possible to insert template guards for context variables. The PR I'm working on changes [the "template context guard" of `NgIf`](https://github.com/angular/angular/blob/cca26166376d4920f5905b168e70ea2e8d70da77/packages/common/src/directives/ng_if.ts#L235) to let the context type be inferred to `NgIfContext<NonNullable<T>>` so that its properties `$implicit` and `ngIf` would no longer include the nullable types.

**Comment by Splaktar** [comment]
Opened https://github.com/angular/angular/issues/34572 to track the issue with ngIf + async pipe type narrowing.

**Comment by Splaktar** [comment]
PR https://github.com/angular/material.angular.io/pull/926 reverts the workarounds, mentioned above, to get the builds working with the async pipe type narrowing bug.

**Comment by Splaktar** [comment]
@jelbourn this is the PR I mentioned in the meeting and you said that you would take a quick look at. This would get us closer to being aligned with the Angular CLI's strict mode.

**Review by andrewseguin — APPROVED** [review]
LGTM

**Pull request merged**